src/core/analyzer-cache-context.ts

```diff
@@ -259,35 +253,99 @@ export class AnalyzerCacheContext {
+          try {
+            const parsedDoc = await this._parse(resolvedUrl, contents);
+            const scannedDocument = await this._scanDocument(parsedDoc);
+
+            // Find all non-lazy imports
+            // TODO(justinfagnani): I think we should scan lazily imported
+            // documents
+            // since we know about them, we should load them. Their features
+            // should
+            // possibly be separated out at export time via better definition of
+            // scopes
+            const imports = scannedDocument.getNestedFeatures().filter(
+                (e) => e instanceof ScannedImport &&
+                    e.type !== 'lazy-html-import') as ScannedImport[];
+
+            // Update dependency graph
+            const importUrls = imports.map((i) => this._resolveUrl(i.url));
+            this._cache.dependencyGraph.addDocument(resolvedUrl, importUrls);
+
+            return scannedDocument;
+          } catch (e) {
+            this._cache.dependencyGraph.rejectDocument(resolvedUrl, e);
+            throw e;
+          }
+        });
+  }
+
+  /**
    * Scan a toplevel document and all of its transitive dependencies.
    */
-  private async _scan(
-      resolvedUrl: string, contents?: string,
-      visited?: Set<string>): Promise<ScannedDocument|'visited'> {
-    if (visited && visited.has(resolvedUrl)) {
-      return 'visited';
-    }
-    const actualVisited = visited || new Set();
-    actualVisited.add(resolvedUrl);
-    const scannedDocument =
-        await this._cache.scannedDocumentPromises.getOrCompute(
-            resolvedUrl, async() => {
-              const parsedDoc = await this._parse(resolvedUrl, contents);
-              return this._scanDocument(parsedDoc, actualVisited);
+  private async _scan(resolvedUrl: string, contents?: string):
+      Promise<ScannedDocument> {
+    return this._cache.dependenciesScannedPromises.getOrCompute(
+        resolvedUrl, async() => {
+          const scannedDocument = await this._scanLocal(resolvedUrl, contents);
+          // Find all non-lazy imports
+          // TODO(justinfagnani): I think we should scan lazily imported
+          // documents
+          // since we know about them, we should load them. Their features
+          // should
+          // possibly be separated out at export time via better definition of
+          // scopes
+          const imports = scannedDocument.getNestedFeatures().filter(
+              (e) => e instanceof ScannedImport &&
+                  e.type !== 'lazy-html-import') as ScannedImport[];
+
+          // Scan imports
+          for (const scannedImport of imports) {
+            const importUrl = this._resolveUrl(scannedImport.url);
+            // Request a scan of `importUrl` but do not wait for the results to
+            // avoid deadlock in the case of cycles. Later we use the
+            // DependencyGraph
+            // to wait for all transitive dependencies to load.
```

Comment:
Style: reflow this comment so that DependencyGraph isn't alone on a line like that?